src/app/code/community/Dh/SeoPagination/Model/Paginator.php

```diff
@@ -1,103 +1,93 @@
     }
-    
-    protected function _getToolbar()
+
+    /**
+     * @return int
+     */
+    private function getLimit()
     {
-        if (is_null($this->_toolbar)) {
-            $this->_toolbar = Mage::app()->getLayout()->createBlock('catalog/product_list_toolbar');
+        if (null === $this->memoizedLimit) {
+            $this->memoizedLimit = (int) $this->getProductListingToolbarBlock()->getLimit();
         }
-        
-        return $this->_toolbar;
+
+        return $this->memoizedLimit;
     }
-    
+
     /**
-     * Initialise the pager and toolbar etc
-     *
      * @return Mage_Page_Block_Html_Pager
      */
-    protected function _getPager()
+    private function getPager()
     {
-        return Mage::app()->getLayout()->createBlock('page/html_pager')
-                ->setLimit($this->_getToolbar()->getLimit())
-                ->setCollection($this->_productCollection);       
+        if (null === $this->memoizedPager) {
+            $productCollection = $this->getProductCollectionBasedOnCurrentCategoryAndLayer();
+            $this->memoizedPager = Mage::app()->getLayout()->createBlock('page/html_pager');
+            $this->memoizedPager->setLimit($this->getLimit())->setCollection($productCollection);
+        }
+
+        return $this->memoizedPager;
     }
-    
+
     /**
-     * Create the next and prev rel links. There are 3 possible scenarios:
-     * 1. Both next and Prev to be output
-     * 2. Only next to be output
-     * 3. Only prev to be output
-     * 
-     * So, decide here what should be output
-     *
-     * @return Dh_SeoPagination_Model_Paginator
+     * @return string
      */
-    public function createLinks()
+    private function getPreviousPageUrl()
     {
-        $pager = $this->_getPager();
-        $numPages = count($pager->getPages());
+        $pager = $this->getPager();
 
-        //Need this to add the links to later on
-        $headBlock = Mage::app()->getLayout()->getBlock('head');
-        
-        //Determine exactly what needs to be output and 
-        //add to the head block
-        if (!$pager->isFirstPage() && !$pager->isLastPage() && $numPages > 2 ) {
-            $headBlock->addLinkRel('prev', $pager->getPreviousPageUrl());
-            $headBlock->addLinkRel('next', $pager->getNextPageUrl());
+        if ($pager->getCurrentPage() > 2) {
+            return $pager->getPreviousPageUrl();
         }
-        elseif($pager->isFirstPage() && $numPages > 1) {
+
+        return $pager->getPagerUrl([$pager->getPageVarName() => null]);
+    }
+
+    public function createLinks()
+    {
+        /** @var Mage_Page_Block_Html_Head $headBlock */
+        $headBlock = Mage::app()->getLayout()->getBlock('head');
+        $pager = $this->getPager();
+
+        $headBlock->removeItem('link_rel', $pager->getPagerUrl([$pager->getPageVarName() => null]));
+        $headBlock->addLinkRel('canonical', $pager->getPagerUrl());

```

Comment:
Additionally fixed a canonical tag bug in Magento. Canonical tag must not link to first page as it is not a superset.